File: gramatica.py
import sys

def check_g1(input_string):
    """L(G1): Números capicúa con 0 y 1"""
    return input_string == input_string[::-1] and all(c in '01' for c in input_string)

def check_g2(input_string):
    """L(G2): a^n b^{n+1} donde n >= 0"""
    # Caso especial para solo "b"
    if input_string == "b":
        return True
    
    # Verificar que solo contiene 'a's y 'b's
    if not all(c in 'ab' for c in input_string):
        return False
    
    # Verificar que todas las 'a's están primero
    if 'ba' in input_string:
        return False
    
    a_count = input_string.count('a')
    b_count = input_string.count('b')
    
    return b_count == a_count + 1

def check_g3(input_string):
    """L(G3): a^n b^{2n+1} donde n >= 1"""
    if not all(c in 'ab' for c in input_string):
        return False
    a_count = input_string.count('a')
    b_count = input_string.count('b')
    return (b_count == 2 * a_count + 1) and (a_count >= 1)

def check_g4(input_string):
    """L(G4): abb o ab"""
    return input_string in ['ab', 'abb']

def check_g5(input_string):
    """L(G5): a(ab)^n b donde n >= 0"""
    if len(input_string) < 2 or input_string[0] != 'a' or input_string[-1] != 'b':
        return False
    if len(input_string) == 2:
        return True  # "ab"
    middle = input_string[1:-1]
    return all(middle[i:i+2] == 'ab' for i in range(0, len(middle), 2))


def get_matching_grammars(input_string):
    """Devuelve lista de gramáticas que cumplen con la cadena"""
    grammars = []
    if check_g1(input_string): grammars.append(1)
    if check_g2(input_string): grammars.append(2)
    if check_g3(input_string): grammars.append(3)
    if check_g4(input_string): grammars.append(4)
    if check_g5(input_string): grammars.append(5)
    return grammars

def main():
    if len(sys.argv) != 2:
        print("Uso: python gramatica.py archivo.txt")
        sys.exit(1)

    try:
        with open(sys.argv[1], 'r') as file:
            lines = [line.strip() for line in file if line.strip()]
            if not lines:
                print("NO acepta")
                return

            # Verificar que cada línea cumpla AL MENOS UNA gramática
            for line in lines:
                if not get_matching_grammars(line):
                    print("NO acepta")
                    return

            print("acepta")

    except FileNotFoundError:
        print(f"Error: Archivo {sys.argv[1]} no encontrado")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: gramatica.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

bool is_palindrome(const char *str) {
    int len = strlen(str);
    for (int i = 0; i < len / 2; i++) {
        if (str[i] != str[len - i - 1]) {
            return false;
        }
    }
    return true;
}

bool check_grammar(int grammar_num, const char *input) {
    int a_count = 0, b_count = 0;
    int len = strlen(input);
    
    switch (grammar_num) {
        case 1: // L(G1): Números capicúa con 0 y 1
            return is_palindrome(input);
            
        case 2: { // L(G2): a^n b^{n+1} donde n >= 0
            bool in_b_section = false;
            for (int i = 0; i < len; i++) {
                if (input[i] == 'a') {
                    if (in_b_section) return false;
                    a_count++;
                } else if (input[i] == 'b') {
                    in_b_section = true;
                    b_count++;
                } else {
                    return false;
                }
            }
            return (b_count == a_count + 1) && (a_count >= 0);
        }
            
        case 3: { // L(G3): a^n b^{3n} donde n >= 1
            bool in_b_section = false;
            for (int i = 0; i < len; i++) {
                if (input[i] == 'a') {
                    if (in_b_section) return false;
                    a_count++;
                } else if (input[i] == 'b') {
                    in_b_section = true;
                    b_count++;
                } else {
                    return false;
                }
            }
            return (b_count == 3 * a_count) && (a_count >= 1);
        }
            
        case 4: // L(G4): abb o ab
            return (strcmp(input, "abb") == 0) || (strcmp(input, "ab") == 0);
            
        case 5: { // L(G5): a(ab)^n b donde n >= 0
            if (len < 2 || input[0] != 'a' || input[len-1] != 'b') {
                return false;
            }
            if (len == 2) return true; // caso "ab"
            
            for (int i = 1; i < len - 1; i += 2) {
                if (i + 1 >= len - 1) return false;
                if (input[i] != 'a' || input[i+1] != 'b') {
                    return false;
                }
            }
            return true;
        }
            
        default:
            return false;
    }
}

int main() {
    FILE *file;
    char filename[20];
    char line[100];
    
    for (int i = 1; i <= 5; i++) {
        snprintf(filename, sizeof(filename), "gramatica_G%d.txt", i);
        file = fopen(filename, "r");
        
        if (file == NULL) {
            printf("Archivo %s no encontrado, omitiendo G%d\n", filename, i);
            continue;
        }
        
        printf("\nProbando gramatica G%d:\n", i);
        while (fgets(line, sizeof(line), file)) {
            // Eliminar el salto de línea
            line[strcspn(line, "\n")] = '\0';
            
            if (check_grammar(i, line)) {
                printf("'%s' - acepta\n", line);
            } else {
                printf("'%s' - NO acepta\n", line);
            }
        }
        fclose(file);
    }
    
    return 0;
}